python/pyarrow/interchange/from_dataframe.py

```diff
@@ -0,0 +1,529 @@
+
+    dict_array = pa.DictionaryArray.from_arrays(indices, dictionary)
+    return dict_array, buffers
+
+
+def datetime_column_to_array(col: ColumnObject) -> tuple[pa.Array, Any]:
+    """
+    Convert a column holding DateTime data to a NumPy array.
+
+    Parameters
+    ----------
+    col : ColumnObject
+
+    Returns
+    -------
+    tuple
+        Tuple of pa.Array holding the data and the memory owner object
+        that keeps the memory alive.
+    """
+    buffers = col.get_buffers()
+    data_buff, data_type = buffers["data"]
+    try:
+        validity_buff, validity_dtype = buffers["validity"]
+    except TypeError:
+        validity_buff = None
+
+    format_str = data_type[2]
+    unit, tz = parse_datetime_format_str(format_str)
+    data_dtype = pa.timestamp(unit, tz=tz)
+
+    # Data buffer
+    data_pa_buffer = pa.foreign_buffer(data_buff.ptr, data_buff.bufsize)
+
+    # Validity buffer
+    validity_pa_buff = validity_buff
+    bytemask = None
+    if validity_pa_buff:
+        validity_pa_buff, bytemask = validity_buffer(validity_buff,
+                                                     validity_dtype,
+                                                     col.describe_null,
+                                                     col.offset)
+
+    # Constructing a pa.Array from data and validity buffer
+    array = pa.Array.from_buffers(
+        data_dtype,
+        col.size(),
+        [validity_pa_buff, data_pa_buffer],
+        offset=col.offset,
+    )
+
+    null_kind, sentinel_val = col.describe_null
+    if null_kind == ColumnNullType.USE_SENTINEL:
+        # pandas iNaT
+        if sentinel_val == -9223372036854775808:
+            bytemask = np.isnan(np.array([str(e) for e in array.to_pylist()],
+                                dtype="datetime64[ns]"))
+        else:
+            raise NotImplementedError(
+                f"Sentinel value {sentinel_val} for datetime is not yet "
+                "supported.")
+
+    # In case a bytemask was constructed with validity_buffer() call
+    # or with sentinel_value then we have to add the mask to the array
+    if bytemask is not None:
+        return pa.array(array.to_pylist(), mask=bytemask), buffers
+    else:
+        return array, buffers
+
+
+def parse_datetime_format_str(format_str):
+    """Parse datetime `format_str` to interpret the `data`."""
+
+    # timestamp 'ts{unit}:tz'
+    timestamp_meta = re.match(r"ts([smun]):(.*)", format_str)
+    if timestamp_meta:
+        unit, tz = timestamp_meta.group(1), timestamp_meta.group(2)
+        if tz != "":
+            raise NotImplementedError("Timezones are not supported yet")
+        if unit != "s":
+            # the format string describes only a first letter of the unit, so
+            # add one extra letter to convert the unit to numpy-style:
+            # 'm' -> 'ms', 'u' -> 'us', 'n' -> 'ns'
+            unit += "s"
+
+        return unit, tz
+
+    # TODO
+    # date 'td{Days/Ms}'
```

Comment:
Currently I implemented the protocol for `timestamp` dtype (for consuming and producing) and not for `date` or `time.` I do plan to test and add support for those data types also (can do in a separate PR).

Reply:
I think it is fine to only support timestamp. I assume the protocol doesn't actually support `date` and `time`, although `date` could be seen as datetime with day resolution (which you can represent with numpy's datetime64 dtype), but I don't know if that is officially allowed by the protocol.

Reply:
Thanks for the feedback. Will leave it as is for now.